Add checks for unsupported characters within image path and output path

Pfleiderer-Adrian opened the pull request:
<!-- The text within this markup is a comment, and is intended to provide
guidelines to open a Pull Request for the ITK repository. This text will not
be part of the Pull Request. -->


<!-- See the CONTRIBUTING (CONTRIBUTING.md) guide. Specifically:

Start ITK commit messages with a standard prefix (and a space):

 * BUG: fix for runtime crash or incorrect result
 * COMP: compiler error or warning fix
 * DOC: documentation change
 * ENH: new functionality
 * PERF: performance improvement
 * STYLE: no logic impact (indentation, comments)
 * WIP: Work In Progress not ready for merge

Provide a short, meaningful message that describes the change you made.

When the PR is based on a single commit, the commit message is usually left as
the PR message.

A reference to a related issue or pull request (https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests)
in your repository. You can automatically
close a related issues using keywords (https://help.github.com/articles/closing-issues-using-keywords/)

@mentions (https://help.github.com/articles/basic-writing-and-formatting-syntax/#mentioning-people-and-teams)
of the person or team responsible for reviewing proposed changes. -->

## Description
I encountered an issue while attempting to load images using the ITK / SimpleITK framework in my Python script. The traceback indicates that the problem arises when attempting to create an IO object for reading an image file that contains Umlauts (ä, ü, ö) in its path. See: #4388

Providing a more descriptive error message indicating that the path cannot be read due to special characters, such as Umlauts, would be helpful for users to understand and address the issue.

## Changes
I have added a simple filepath check for illegal characters in the _NewImageReader and imwrite routines. Additionally, for the imwrite routine, a check was added to verify if the output directory exists.

## PR Checklist
- [x] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [x] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)

<!-- **Thanks for contributing to ITK!** -->

dzenanz commented on Wrapping/Generators/Python/itk/support/extras.py:
```diff
@@ -1165,6 +1165,24 @@ def imwrite(
+        try:
+            filename.encode('ascii')
+        except UnicodeEncodeError:
+            msg += "\nThe output path contains not supported special characters. \n" + f"Filename = {filename}"
```
not supported -> unsupported

Perhaps give user a hint about setting UTF-8 as the default code page?

blowekamp replied:
It may be useful to place this check into a function.

dzenanz commented on Wrapping/Generators/Python/itk/support/template_class.py:
```diff
@@ -157,6 +158,12 @@ def firstIfList(arg):
+                try:
+                    inputFileName.encode('ascii')
+                except UnicodeEncodeError:
+                    msg += "\nThe image path contains not supported special characters. \n" + f"Filename = {inputFileName}"
```
Preferably give a hint about UTF8 code page.

dzenanz reviewed (commented):
Ah, this check in makes sense via Python interface. Mostly looks good.

blowekamp commented:
There may be something in the SWIG layer that can be down with the conversion. There is some information here:
https://www.swig.org/Doc3.0/Python.html#Python_nn77

That needs to be gone through.

jhlegarreta commented:
@Pfleiderer-Adrian thanks for having a look at this. A few comments:
- Commit 7304267 does not addess a bug, but a `STYLE` change.
- Please, add a meaningful commit message body in line with [ITK commit message practices](https://docs.itk.org/en/latest/contributing/index.html#commit-messages).

dzenanz commented:
The second commit should be squashed into the first one.

blowekamp commented on Wrapping/Generators/Python/itk/support/extras.py:
```diff
@@ -1165,6 +1165,24 @@ def imwrite(
+        msg += "\nThe output dir doesn't exist. \n" + f"Filename = {filename}"
+        raise RuntimeError(
+            f"Could not create IO object for writing file {filename}" + msg
+        )
```
This would be a useful check to occur in the C++ code.

dzenanz replied:
I agree, but I assume that performing an equivalent check in C++ would be a lot harder. If the author has the will to explore this, it would be good.

blowekamp reviewed (commented):
A couple suggestions.

Pfleiderer-Adrian commented:
Thank you for your feedback! I'll post an update this weekend. :)

N-Dekker commented:
Thank you for diving into this issue, @Pfleiderer-Adrian  Do I understand correctly that this issue (#4388) is only a problem for _Windows_ users, not for Linux users? (Specifically, only those Windows users that do not have UTF-8 support enabled _and_ selected a locale that does not support the filename?) I'm asking, because it feels too strict to _only_ accept ASCII characters from now, for all users.

So is it really the intention to reject any non-ASCII character in file names, for any user? Aren't we then throwing the baby out with the bath water?

Would it be an idea to _only_ do those checks _after_ trying to use `imageIO`, and _only_ when `imageIO` has really failed to do its job?

hjmjohnson committed:
BUG: Add checks for special characters in file paths

- In `imwrite` function, add validation to ensure the output directory exists,
  raising an error if not.
- Check for unsupported special characters in file paths, raising an error if
  present.
- Apply similar checks to `itkTemplate` to validate input file paths for existence and encoding.

hjmjohnson reviewed (changes requested):
This comment should be addressed before this is finalized.

https://github.com/InsightSoftwareConsortium/ITK/pull/4390#issuecomment-1888819346

dzenanz reviewed (commented):
This makes the Python wrapping behave differently from C++.

hjmjohnson commented:
@Pfleiderer-Adrian thank you for surfacing this issue and starting the conversation — your PR put a real spotlight on the Windows non-ASCII-filename failure mode, and the reviewer discussion that followed (esp. @N-Dekker's pointer to [SuperElastix/elastix#1401](https://github.com/SuperElastix/elastix/pull/1401) by @codeling, and the @blowekamp / @dzenanz preference for fixing the C++ side rather than tightening the Python contract) shaped the takeover direction.

I've opened **#6231** as a successor that follows the elastix recipe: an 8-line UTF-8 active-code-page manifest auto-attached to every ITK test executable on Windows. With Windows 10 1903+, this lets the existing `const char *` filename APIs ingest UTF-8 byte sequences without code changes — no need to reject anything in the Python wrapper. You're credited as `Co-Authored-By:` on the PR commit.

A library-side follow-up (Tier B) will wire ITK's existing-but-unused `itkInternationalizationIOHelpers` (`i18n_open` / `i18n_fopen`) through `ImageIOBase` / PNG / JPEG so SimpleITK and Python users (where the manifest doesn't reach `python.exe`) also benefit. That'll be a separate PR, also co-authored with you.

Suggest closing this PR after #6231 merges. Thanks again for the contribution.

Pfleiderer-Adrian commented:
@hjmjohnson Thanks a lot for the detailed follow-up and for carrying this forward — really appreciate the co-author credit! The approach in #6231 sounds great to me. Happy to successfull close this PR.
